### PR TITLE
Integration with 3GPP 26.973 v15.1.0 - New operators

### PR DESCRIPTION
**Pull request by ludomal**
Refers to #85.

These commits provide the code for the new operators defined in [3GPP 26.973](https://portal.3gpp.org/desktopmodules/Specifications/SpecificationDetails.aspx?specificationId=3326).

### CODE REVIEW

File: src/basop/control.h [hunk 1]
```diff
@@ -85,10 +98,10 @@ static __inline void incrFor (void) {
  *
  *****************************************************************************/
 #ifndef WMOPS
-#define WHILE( a) while( a)
+#define WHILE(a) while (a)
```
Comment (dennisguse) [inline src/basop/control.h]: Really?

File: src/basop/control.h [hunk 2]
```diff
@@ -137,13 +150,16 @@ static __inline void incrWhile (void) {
  *
  *****************************************************************************/
 #ifndef WMOPS
-#define IF( a) if( a)
+#define IF (a) if (a)
```
Comment (dennisguse) [inline src/basop/control.h]: Really?

**Comment by dennisguse** [comment]
@ludomal Is the WMOPS performance measurement really necessary?
I mean it is working nicely while from code maintenance perspective it is rather complicated...

Can we at least remove the `GOTO`?

File: src/basop/control.h (hunk 1 above)
Comment (ludomal) [reply src/basop/control.h]: Yeah... There are a few instances where white spaces were adjusted. I did not push files that had only changes in white spaces, but I left the changes in files for which contents were adjusted too. I assumed we will run a script to normalize the code later?

**Comment by dennisguse** [comment]
@ludomal Beside the GOTO (a personal thing :) ), the code looks okay.
Would you mind adding a prefix to the commit messages?
Something like "3GPP:" would make it easier to find these commits later.

**Comment by dennisguse** [comment]
At least the GOTO (also WHILE) is not used ;)
As it only occurs three times.

@ludomal Sounds like a plan, then let's merge and schedule maintenance work for later (saves time now).

```
user@noname:~/hack/STL$ git checkout remotes/origin/pr/105
Note: checking out 'remotes/origin/pr/105'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at 68b092e... Adds definition for Word64
user@noname:~/hack/STL$ cd src/
user@noname:~/hack/STL/src$ ag GOTO
basop/count.c
101:  "L40_round",     "L_saturate40",    "round40",       "IF",              "GOTO",

basop/basop_cnt.c
101:  "GOTO",

basop/control.h
277: *  Function Name : GOTO
281: *    The macro GOTO should be used instead of the 'goto' C statement.
287:#define GOTO goto
290:#define GOTO if (incrGoto (), 0); else goto
user@noname:~/hack/STL/src$ 

```

File: src/basop/count.c [hunk 3]
```diff
@@ -77,58 +77,238 @@ void setFrameRate (int samplingFreq, int frameLength) {
...
+  1,     1,     1,     1,     1,
+  1,     1,     1,     1,     1,
+  1,     1,     2,     2,     1,
+  1,     1,     1,     2,     1,
+
+  1,     1,     1,     2,     1,
+  1,     1,     18,    1,     1,
+  1,     1,     1,     1,     1,
+  1,     1,     1,     1,     1,
+  2,     2,     2,     2,     1,
```
Comment (dennisguse) [inline src/basop/count.c]: Here the weights are changed - any reason why?

File: src/basop/count.c [hunk 4]
```diff
@@ -159,14 +339,14 @@ static Word32 LastWOper[MAXCOUNTERS];
 
 #ifdef WMOPS
 static char *my_strdup (const char *s) {
```
Comment (dennisguse) [inline src/basop/count.c]: Function is not used?

```
user@noname:~/hack/STL/src$ ag my_strdup
basop/count.c
341:static char *my_strdup (const char *s) {
369:  objectName[++maxCounter] = my_strdup (objectNameArg);
user@noname:~/hack/STL/src$ 
```

File: src/basop/count.h [hunk 5]
```diff
@@ -265,7 +265,7 @@ typedef struct {
   UWord32 L40_msu;              /* Complexity Weight of 1 */
   UWord32 msu_r40;              /* Complexity Weight of 2 */
   UWord32 Mpy_32_16_ss;         /* Complexity Weight of 2 */
-  UWord32 Mpy_32_32_ss;         /* Complexity Weight of 4 */
+  UWord32 Mpy_32_32_ss;         /* Complexity Weight of 2 */
```
Comment (dennisguse) [inline src/basop/count.h]: Any reason?

File: src/basop/count.c (hunk 3 above)
Comment (ludomal) [reply src/basop/count.c]: @dennisguse, the weights were set 10 years ago and DSPs have changed. The old weights were discussed amongst 3GPP experts and were updated to reflect the modern DSP architectures.

**Comment by ludomal** [comment]
@dennisguse ,these tools are used to instrument codecs to simulate DSPs and evaluate the complexity of the algorithms. They are ment to be included in other projects. 
When algorithms are executed, it counts the number of call to each instruction (the number of goto and while instances for example). Then the weights help estimate the complexity of the algorithm.

That's why I proposed we ask users of these operators for guidance on what could be excluded.

**Comment by ludomal** [comment]
@dennisguse , are we good to merge this pull request?

I have been trying to rebase to add [3GPP] in front of each comment but the LaTex/graphics/lp* issue gets in the way with all methods I have tried. Is it possible to edit the comments from the GitHub interface?

**Review by dennisguse — APPROVED** [review]
Let's go.

**Pull request merged**